design/open/spark-outputcommitter/committer.md

```diff
@@ -179,10 +263,10 @@ Now change LakeFSOutputCommitter to loop during HOC commit:
 	use a new "merge but *never* copy from source branch" strategy).
   - Go back and attempt another merge.
 
-This is essentially (noncooperative!) locking of output paths on top of
-lakeFS, with no additional DB.  We can even add cooperation by means of
-various locking hints, _informing_ multiple jobs about attempting to update
-the same paths but keeping things safe regardless.
+This is essentially (noncooperative but optimistic!) locking of output paths
```

Comment:
To clarify, are you suggesting a more sophisticated merge-if here? I mean, a merge-if logic that not only checks whether the destination branch changes but what was changes on that branch?

Reply:
This is the merge-if operation defined on l. 253.  It returns only success or failure; if it fails, the caller can decide to diagnose the differences if it wants to try again.  But in fact, here we shall directly merge the destination into the source and try again.  I am not sure how listing the changes on the branch would help here.  It seems slower because we would still have to merge, and it would be unreliable because with separate calls there would still be a race between diffing and merging.